server/implementation/src/main/java/io/smallrye/graphql/bootstrap/Bootstrap.java

```diff
@@ -846,70 +881,67 @@ private GraphQLInputObjectField createGraphQLInputObjectFieldFromField(Field fie
 
         // Default value (on method)
         if (field.hasDefaultValue()) {
-            inputFieldBuilder = inputFieldBuilder.defaultValue(sanitizeDefaultValue(field));
+            inputFieldBuilder = inputFieldBuilder.defaultValueProgrammatic(sanitizeDefaultValue(field));
         }
 
         return inputFieldBuilder.build();
     }
 
     private GraphQLInputType createGraphQLInputType(Field field) {
-
         GraphQLInputType graphQLInputType = referenceGraphQLInputType(field);
 
-        Wrapper wrapper = dataFetcherFactory.unwrap(field, false);
+        // Check if field is mandatory
+        if (field.isNotNull()) {
+            graphQLInputType = GraphQLNonNull.nonNull(graphQLInputType);
+        }
 
-        // Collection
+        Wrapper wrapper = dataFetcherFactory.unwrap(field, false);
+        // Field can have a wrapper, like List<String>
         if (wrapper != null && wrapper.isCollectionOrArrayOrMap()) {
-            // Mandatory in the collection
-            if (wrapper.isNotEmpty()) {
-                graphQLInputType = GraphQLNonNull.nonNull(graphQLInputType);
-            }
-            // Collection depth
+            // Loop as long as there is a wrapper
             do {
                 if (wrapper.isCollectionOrArrayOrMap()) {
                     graphQLInputType = list(graphQLInputType);
+                    // Wrapper itself can also be mandatory
+                    if (wrapper.isNotEmpty()) {
+                        graphQLInputType = GraphQLNonNull.nonNull(graphQLInputType);
```

Comment:
Actually I think the `wrapper.isNotEmpty()` method has a rather confusing name and it should instead be `isWrappedTypeNotNull()` - it refers to non-nullability of the wrapped type, not the wrapper itself. Therefore moving this `graphQLInputType = GraphQLNonNull.nonNull(graphQLInputType);` after `graphQLInputType = list(graphQLInputType)` is wrong because then you mark the OUTER type (the wrapper) as non-null based on the INNER type being non-null. 

I think we should rename `isNotEmpty` to `isWrappedTypeNotNull` and then here we would have this method as follows:
```
private GraphQLInputType createGraphQLInputType(Field field) {
        GraphQLInputType graphQLInputType = referenceGraphQLInputType(field);

        Wrapper wrapper = dataFetcherFactory.unwrap(field, false);
        // Field can have a wrapper, like List<String>
        if (wrapper != null && wrapper.isCollectionOrArrayOrMap()) {
            // Loop as long as there is a wrapper
            do {
                if (wrapper.isCollectionOrArrayOrMap()) {
                    if (wrapper.isWrappedTypeNotNull()) {
                        graphQLInputType = GraphQLNonNull.nonNull(graphQLInputType);
                    }
                    graphQLInputType = list(graphQLInputType);
                    wrapper = wrapper.getWrapper();
                } else {
                    wrapper = null;
                }
            } while (wrapper != null);
        }

        // Check if field is mandatory
        if (field.isNotNull()) {
            graphQLInputType = GraphQLNonNull.nonNull(graphQLInputType);
        }

        return graphQLInputType;
    }
``` 
??? The tests pass for me then. But I may not properly understand the reason for your changes here.

AFAIU there's no concept of not-empty collections in GraphQL. If the inner type is marked as not-null, it means that the collection must not contain null entries, NOT that it must not be empty.

Reply:
Maybe @phillip-kruger remembers details about the `wrapper.isNotEmpty` method and can correct me, but I think it means that the wrapped type is not-null

Reply:
That makes a lot more sense, thank you for the explanation. I implemented the fixes you mentioned and now the tests pass successfully.
https://github.com/smallrye/smallrye-graphql/pull/2010/commits/f0bd7447f402cc9eb0d8696ffbfad7c4422db107